Support array column type

instcode opened the pull request:
This PR implements `array` column type support for Postgres.

The implementation supports converting both single/multiple dimension array to string & json. However, in json case, only simple value types (number, bool, and text) are allowed in the array elements, e.g. integer[], text[][], bigint[][][]...

The implementation does have a compatible limitation. It works perfectly for Postgres server version 8.3.0 and above due to the usage of object type versus primitive type for array elements. This can be fixed but it would introduce a very complex code base.

instcode commented:
Hi @hito4t, could you please have a look? Thank you.

hito4t commented on embulk-input-jdbc/src/main/java/org/embulk/input/jdbc/getter/ArrayColumnGetter.java:
```diff
@@ -0,0 +1,103 @@
+    @Override
+    public void stringColumn(Column column)
+    {
+        to.setString(column, value.toString());
```
java.sql.Array is interface and doesn't implement toString method.
We should implement the code converting Array to String by hand, otherwise the result will vary depends on DBMS.

instcode replied:
Yes, because the implementation is kinda specific for Postgres, I refactored the code from `embulk-input-jdbc` to `embulk-input-postgresql`.

hito4t commented on embulk-input-jdbc/src/main/java/org/embulk/input/jdbc/getter/ColumnGetterFactory.java:
```diff
@@ -62,6 +62,8 @@ private ColumnGetter newColumnGetter(JdbcInputConnection con, PluginTask task, J
             return new TimestampColumnGetter(to, toType, newTimestampFormatter(option, DateColumnGetter.DEFAULT_FORMAT));
         case "decimal":
             return new BigDecimalColumnGetter(to, toType);
+        case "array":
```
Because "array" type is specific for PostgreSQL, these lines should be in PostgreSQLColumnGetterFactory class.

instcode replied:
Got it. I moved to `embulk-input-postgresql`.

instcode commented:
Thanks @hito4t for your review comments. I refactored the code and added unit-test to handle the `array` type specific for Postgres. I also updated the README to add more information about this support. Please have a look ^^ Thanks.

instcode commented:
BTW, the tests run okay on my machine but it seems Travis says no :-). The build stops at `embulk-input-mysql` errors though.

sakama commented:
@instcode I think Travis CI failure will be fixed by https://github.com/embulk/embulk-input-jdbc/pull/122

hito4t commented:
@instcode
Great!
Thank you!
